types.go

```diff
@@ -41,6 +43,159 @@ func (o Strings) MarshalJSON() ([]byte, error) {
 var _ json.Unmarshaler = (*Strings)(nil)
 var _ json.Marshaler = (*Strings)(nil)
 
+// Flag represents a ternary value: false (-1), default (0), or true (+1).
+//
+// When encoded in json, False is "false", Default is "null" (or empty), and True
+// is "true".
+type Flag int8
+
+const (
+	False   Flag = -1
+	Default Flag = 0
+	True    Flag = 1
+)
+
+// WithDefault resolves the value of the flag given the provided default value.
+func (f Flag) WithDefault(defaultValue bool) bool {
+	switch f {
+	case False:
+		return false
+	case Default:
+		return defaultValue
+	case True:
+		return true
+	default:
+		panic(fmt.Sprintf("invalid flag value %d", f))
+	}
+}
+
+func (f Flag) MarshalJSON() ([]byte, error) {
+	switch f {
+	case Default:
+		return json.Marshal(nil)
+	case True:
+		return json.Marshal(true)
+	case False:
+		return json.Marshal(false)
+	default:
+		return nil, fmt.Errorf("invalid flag value: %d", f)
+	}
+}
+
+func (f *Flag) UnmarshalJSON(input []byte) error {
+	switch string(input) {
+	case "null", "undefined":
+		*f = Default
+	case "false":
+		*f = False
+	case "true":
+		*f = True
+	default:
+		return fmt.Errorf("failed to unmarshal %q into a flag: must be null/undefined, true, or false", string(input))
+	}
+	return nil
+}
+
+func (f Flag) String() string {
+	switch f {
+	case Default:
+		return "default"
+	case True:
+		return "true"
+	case False:
+		return "false"
+	default:
+		return fmt.Sprintf("<invalid flag value %d>", f)
+	}
+}
+
+var _ json.Unmarshaler = (*Flag)(nil)
+var _ json.Marshaler = (*Flag)(nil)
+
+// Priority represents a value with a priority where 0 means "default" and -1
+// means "disabled".
+//
+// When encoded in json, Default is encoded as "null" and Disabled is encoded as
+// "false".
+type Priority int64
+
+const (
+	DefaultPriority Priority = 0
+	Disabled        Priority = -1
+)
+
+// WithDefault resolves the priority with the given default.
+//
+// If `defaultPriority` is Default/0, this function will return 0.
+func (p Priority) WithDefault(defaultPriority Priority) (priority int64, enabled bool) {
+	switch p {
+	case Disabled:
+		return 0, false
+	case DefaultPriority:
+		if defaultPriority < 0 {
```

Comment:
Should we be checking defaultPriority == Disabled? This seems to imply that a `Priority` is allowed to be `<-1` which is invalid

Reply:
Eh, I'd rather not add the additional error case. (there's a theme here)